Setup Script

prayanshchh opened the pull request:
<!--
This section can be deleted after reading.

We employ the following branching strategy to simplify the development process and to ensure that only stable code is pushed to the `master` branch:

- `develop`: For unstable code: New features and bug fixes.
- `master`: Where the stable production ready code lies. Only security related bugs.

NOTE!!!

ONLY SUBMIT PRS AGAINST OUR `DEVELOP` BRANCH. THE DEFAULT IS `MAIN`, SO YOU WILL HAVE TO MODIFY THIS BEFORE SUBMITTING YOUR PR FOR REVIEW. PRS MADE AGAINST `MAIN` WILL BE CLOSED.

-->

<!--
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request.
-->

**What kind of change does this PR introduce?**
created setup script for `develop-postgres`

<!-- E.g. a bugfix, feature, refactoring, etc… -->

**Issue Number:**
#2688 

**Snapshots/Videos:**

[Screencast from 2025-01-29 08-47-05.webm](https://github.com/user-attachments/assets/2e4ad7f8-14ae-4f1d-af19-8dda98de59cb)

[Screencast from 2025-01-29 08-52-29.webm](https://github.com/user-attachments/assets/7af3137c-b34e-4976-adb7-5d18310053b7)

[Screencast from 2025-01-29 08-54-58.webm](https://github.com/user-attachments/assets/4a8630ba-1ecf-4c0e-bc3d-b030f6158f16)


<!--Add snapshots or videos wherever possible.-->

**If relevant, did you update the documentation?**

<!--Add link to Talawa-Docs.-->

**Summary**

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

**Does this PR introduce a breaking change?**

<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

## Checklist

### CodeRabbit AI Review
- [x] I have reviewed and addressed all critical issues flagged by CodeRabbit AI
- [x] I have implemented or provided justification for each non-critical suggestion
- [x] I have documented my reasoning in the PR comments where CodeRabbit AI suggestions were not implemented

### Test Coverage
- [x] I have written tests for all new changes/features
- [x] I have verified that test coverage meets or exceeds 95%
- [x] I have run the test suite locally and all tests pass


**Other information**

<!--Add extra information about this PR here-->

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

<!--Yes or No-->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced an interactive setup process that streamlines the configuration of environment variables and application settings for both local and CI environments.
	- Enhanced input validations and error handling to provide clearer, more reliable feedback during setup.

- **Documentation**
	- Updated the installation guide with detailed, step-by-step instructions for running the new setup script and configuring the environment.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

prayanshchh commented:
will fix failing tests and implement suggestions

palisadoes commented:
1. The documentation says that a JWT needs to be generated. Is this included in your code?
2. Why does the prompt for production or development come last? It should be first, because the environment variables will be set differently afterwards

prayanshchh commented:
> 1. The documentation says that a JWT needs to be generated. Is this included in your code?
> 2. Why does the prompt for production or development come last? It should be first, because the environment variables will be set differently afterwards

will implement both

prayanshchh commented:
> 1. The documentation says that a JWT needs to be generated. Is this included in your code?
> 2. Why does the prompt for production or development come last? It should be first, because the environment variables will be set differently afterwards

`NODE_ENV` is not the deciding factor as to whether test env should be configured or devcontainer, the variable `CI` is the deciding factor, if set false  `.env.devcontainer` will be referred, if true, .env.ci will be referred to fill the .env.

![image](https://github.com/user-attachments/assets/710ce787-96c5-4bb1-b6a0-94035a8d79fd)
![image](https://github.com/user-attachments/assets/447b1ff3-5945-45ab-aba9-e8ccb561df1c)
![image](https://github.com/user-attachments/assets/eb40dd7b-8365-41e0-8db0-29f8de636741)

I am planning to go with this approach, is this alright?

palisadoes commented:
@xoldd PTAL at this PR. It has implications for API setup using a script

palisadoes commented:
> > 1. The documentation says that a JWT needs to be generated. Is this included in your code?
> > 2. Why does the prompt for production or development come last? It should be first, because the environment variables will be set differently afterwards
> 
> `NODE_ENV` is not the deciding factor as to whether test env should be configured or devcontainer, the variable `CI` is the deciding factor, if set false `.env.devcontainer` will be referred, if true, `.env.ci` will be referred to fill the `.env`.
> 
> ![image](https://private-user-images.githubusercontent.com/135210710/407191377-710ce787-96c5-4bb1-b6a0-94035a8d79fd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgwNDIwMDMsIm5iZiI6MTczODA0MTcwMywicGF0aCI6Ii8xMzUyMTA3MTAvNDA3MTkxMzc3LTcxMGNlNzg3LTk2YzUtNGJiMS1iNmEwLTk0MDM1YThkNzlmZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyOFQwNTIxNDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZmFmOGQ0ZGI4OGE3ZTJlYThkMDg1MDAxNDgxMjljODM2NGRmMjcyMDQxZTc2OGE4MGI3ZjdlMzlkMWRiMWQ3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.-fUWMsY8Hvjb_EDB3qwknqAX96L8QR-ITlLErlBDpDY) ![image](https://private-user-images.githubusercontent.com/135210710/407191422-447b1ff3-5945-45ab-aba9-e8ccb561df1c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgwNDIwMDMsIm5iZiI6MTczODA0MTcwMywicGF0aCI6Ii8xMzUyMTA3MTAvNDA3MTkxNDIyLTQ0N2IxZmYzLTU5NDUtNDVhYi1hYmE5LWU4Y2NiNTYxZGYxYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyOFQwNTIxNDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wNzlhZGU5MTBmMDllYjQxMjA0MWEyNjk3MDk0NGYzMTY2MTlmMTExYTNkYTk1YmY1YTBhYTdiNzc0YTFkNWJmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.917n-dSwoDy_PpQ1eQWB5nSOAxBlGgPUzrtEZZ7Sw4o) ![image](https://private-user-images.githubusercontent.com/135210710/407191467-eb40dd7b-8365-41e0-8db0-29f8de636741.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgwNDIwMDMsIm5iZiI6MTczODA0MTcwMywicGF0aCI6Ii8xMzUyMTA3MTAvNDA3MTkxNDY3LWViNDBkZDdiLTgzNjUtNDFlMC04ZGIwLTI5ZjhkZTYzNjc0MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyOFQwNTIxNDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00ZDQ3ZDg4MmJmMTY5Y2YwMmFmMjZkY2UyNTU3Y2ZkZDc2MThhM2FjNDljMmNlMTRjYzdhMGI2YjhjNmQ4Zjc4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.DFZAz2pIX4XLc_kR3RrfkaP0sW0HR5HADQ4WEzlRfvE)
> 
> I am planning to go with this approach, is this alright?

1. It looks that way. Please go ahead.
2. What about the DB importation?

prayanshchh commented:
> > > 1. The documentation says that a JWT needs to be generated. Is this included in your code?
> > > 2. Why does the prompt for production or development come last? It should be first, because the environment variables will be set differently afterwards
> > 
> > 
> > `NODE_ENV` is not the deciding factor as to whether test env should be configured or devcontainer, the variable `CI` is the deciding factor, if set false `.env.devcontainer` will be referred, if true, `.env.ci` will be referred to fill the `.env`.
> > ![image](https://private-user-images.githubusercontent.com/135210710/407191377-710ce787-96c5-4bb1-b6a0-94035a8d79fd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgwNDIwMDMsIm5iZiI6MTczODA0MTcwMywicGF0aCI6Ii8xMzUyMTA3MTAvNDA3MTkxMzc3LTcxMGNlNzg3LTk2YzUtNGJiMS1iNmEwLTk0MDM1YThkNzlmZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyOFQwNTIxNDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ZmFmOGQ0ZGI4OGE3ZTJlYThkMDg1MDAxNDgxMjljODM2NGRmMjcyMDQxZTc2OGE4MGI3ZjdlMzlkMWRiMWQ3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.-fUWMsY8Hvjb_EDB3qwknqAX96L8QR-ITlLErlBDpDY) ![image](https://private-user-images.githubusercontent.com/135210710/407191422-447b1ff3-5945-45ab-aba9-e8ccb561df1c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgwNDIwMDMsIm5iZiI6MTczODA0MTcwMywicGF0aCI6Ii8xMzUyMTA3MTAvNDA3MTkxNDIyLTQ0N2IxZmYzLTU5NDUtNDVhYi1hYmE5LWU4Y2NiNTYxZGYxYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyOFQwNTIxNDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wNzlhZGU5MTBmMDllYjQxMjA0MWEyNjk3MDk0NGYzMTY2MTlmMTExYTNkYTk1YmY1YTBhYTdiNzc0YTFkNWJmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.917n-dSwoDy_PpQ1eQWB5nSOAxBlGgPUzrtEZZ7Sw4o) ![image](https://private-user-images.githubusercontent.com/135210710/407191467-eb40dd7b-8365-41e0-8db0-29f8de636741.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgwNDIwMDMsIm5iZiI6MTczODA0MTcwMywicGF0aCI6Ii8xMzUyMTA3MTAvNDA3MTkxNDY3LWViNDBkZDdiLTgzNjUtNDFlMC04ZGIwLTI5ZjhkZTYzNjc0MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyOFQwNTIxNDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00ZDQ3ZDg4MmJmMTY5Y2YwMmFmMjZkY2UyNTU3Y2ZkZDc2MThhM2FjNDljMmNlMTRjYzdhMGI2YjhjNmQ4Zjc4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.DFZAz2pIX4XLc_kR3RrfkaP0sW0HR5HADQ4WEzlRfvE)
> > I am planning to go with this approach, is this alright?
> 
> 1. It looks that way. Please go ahead.
> 2. What about the DB importation?

@gurramkarthiknetha is working on it, we talked on slack, he has started work on this

prayanshchh commented on setup.ts:
```diff
@@ -0,0 +1,448 @@
+			name: "useDefaultMinio",
+			message:
+				"Do you want to use the recommended default Minio settings? (Y)/N",
+			default: true,
+		},
+	]);
+	if (!useDefaultMinio) {
+		await minioSetup();
+	}
+
+	const { useDefaultCloudbeaver } = await inquirer.prompt([
+		{
+			type: "confirm",
+			name: "useDefaultCloudbeaver",
+			message:
+				"Do you want to use the recommended default CloudBeaver settings? (Y)/N",
+			default: true,
+		},
+	]);
+	if (!useDefaultCloudbeaver) {
+		await cloudbeaverSetup();
+	}
+
+	const { useDefaultPostgres } = await inquirer.prompt([
+		{
+			type: "confirm",
+			name: "useDefaultPostgres",
+			message:
+				"Do you want to use the recommended default Postgres settings? (Y)/N",
+			default: true,
+		},
+	]);
+	if (!useDefaultPostgres) {
+		await postgresSetup();
+	}
+	await setNodeEnvironment();
+	await administratorEmail();
+
+	console.log("Configuration complete.");
+}
```
shall I provide a summary at the end @palisadoes

palisadoes commented:
@coderabbitai full review

prayanshchh commented:
sir this isn't ready as of moment, I have implemented you requirements, just fixing code_quality issues, then I can look at suggestions all together

prayanshchh commented:
will implement all the new suggestions the old ones are not relevant as a lot of code was changed

palisadoes commented:
1. You need to update the video. You said the prod/dev question would come first.
2. What happens when non defaults are selected?
3. When is the .env file updated? It needs to be only at the very end.
4. What happens if a .env file is already discovered?

prayanshchh commented:
1. I have updated and uploaded three videos
2. the video shows how the script handles non-defaults
3. presently we update env file as the user enters the values, I have implemented backup and restoring env, so if ctrl+c is hit then env goes back to its previous state, is this good?
4. I will add this functionality

prayanshchh commented on src/setup.ts:
```diff
@@ -0,0 +1,674 @@
+	]);
+	updateEnvVariable({ API_JWT_EXPIRES_IN });
+
+	const jwtSecret = generateJwtSecret();
+
+	const { API_JWT_SECRET } = await inquirer.prompt([
+		{
+			type: "input",
+			name: "API_JWT_SECRET",
+			message: "JWT secret:",
+			default: jwtSecret,
+			validate: (input: string) => {
+				if (input.length < 64) {
+					return "JWT secret must be at least 64 characters long.";
+				}
+				return true;
+			},
+		},
+	]);
+	updateEnvVariable({ API_JWT_SECRET });
```
do we need this @palisadoes ?

prayanshchh commented on test/setup/setNodeEnvironment.test.ts:
```diff
@@ -0,0 +1,47 @@
+		process.env.NODE_ENV = originalNodeEnv;
+		vi.resetAllMocks();
+	});
+	it("should update NODE_ENV when selection is made", async () => {
+		const mockedNodeEnv = "development";
+		vi.spyOn(inquirer, "prompt").mockResolvedValueOnce({
+			NODE_ENV: mockedNodeEnv,
+		});
+		await setNodeEnvironment();
+
+		expect(process.env.NODE_ENV).toBe(mockedNodeEnv);
+	});
+
+	it("should prompt the user for NODE_ENV when NODE_ENV is 'test'", async () => {
+		const mockedNodeEnv = "production";
+		vi.spyOn(inquirer, "prompt").mockResolvedValueOnce({
+			NODE_ENV: mockedNodeEnv,
+		});
+		process.env.NODE_ENV = "test";
+
+		await setNodeEnvironment();
+
+		expect(process.env.NODE_ENV).toBe(mockedNodeEnv);
+	});
+
+	it("should set NODE_ENV to 'development' by default when NODE_ENV is 'test' and no selection is made", async () => {
+		vi.spyOn(inquirer, "prompt").mockResolvedValueOnce({
+			NODE_ENV: "development",
+		});
+
+		process.env.NODE_ENV = "test";
+
+		await setNodeEnvironment();
+
+		expect(process.env.NODE_ENV).toBe("development");
+	});
```
NODE_ENV setup uses type list of inquirer, user can only enter based on the choices given, so there is no scope of user entering invalid values..also validate attribute is not applicable in list type in inquirer so can't implement validate functionality here

prayanshchh commented:
will implement 3 tommorrow

palisadoes commented:
Please coordinate with @rahulch07 on getting a DB dump from his PR for the setup script

rahulch07 commented:
@prayanshchh sent the dump file. please check slack

prayanshchh commented:
will make the changes now, when the sample data functionality is completed, presently all the four conditions have been met

> 1. You need to update the video. You said the prod/dev question would come first.
> 2. What happens when non defaults are selected?
> 3. When is the .env file updated? It needs to be only at the very end.
> 4. What happens if a .env file is already discovered?

palisadoes reviewed (changes requested):
1. Can you move src/setup.ts to src/setup and write tests for it?
2. We need to have complete coverage. There has been a history of poor setup experiences in the repo

palisadoes commented:
@prayanshchh The DB PR is merged
- https://github.com/PalisadoesFoundation/talawa-api/pull/3118

prayanshchh commented:
> 1. Can you move src/setup.ts to src/setup and write tests for it?
> 2. We need to have complete coverage. There has been a history of poor setup experiences in the repo

yes sir, the tests are already present I will try and bring patch to 100%

palisadoes commented:
Please:

1. fix the failing tests
2. make coderabbit.ai approve your changes

palisadoes reviewed (changes requested):
Please update the database JSON files

1. They should be in a `sample_data` folder not `scripts/`
2. Update `src/utilities/loadSampleData.ts` accordingly

prayanshchh commented:
alright sir I will make the changes, give me some time I have exams going on

prayanshchh commented:
sir I have one doubt on the working of setup script

previously setup script would start the development server hence importing sample data would make sense as database was up and running

but now that we setup the project using devcontainers, when `npm run setup` will be run, the database will not be up and hence if we give option to import sample data it won't be able to import, as postgres will start with devcontainer and not with setup script.

thoughts? @palisadoes

prayanshchh commented:
another thing we don't have a `npm run import sample_data` as we used to have in develop, is this okay?

palisadoes commented:
> another thing we don't have a `npm run import sample_data` as we used to have in develop, is this okay?

Yes

prayanshchh commented:
> > sir I have one doubt on the working of setup script
> > previously setup script would start the development server hence importing sample data would make sense as database was up and running
> > but now that we setup the project using devcontainers, when `npm run setup` will be run, the database will not be up and hence if we give option to import sample data it won't be able to import, as postgres will start with devcontainer and not with setup script.
> > thoughts? @palisadoes
> 
> In that case we'll need 2 scripts. One for setup and the other for DB importation.

alright then I will create another script for db importation, the initial idea is to set the db importation and format db value in env as YES or NO, and when devcontainers setup is running we can run the db importation script, is this good?

I will continue in this PR

palisadoes commented:
OK

1. It will also need to work with production
2. There will need to be precautions that re-importation doesn't happen with each restart

palisadoes commented:
1. Please fix the conflicting files
2. The PR GitHub actions have been fixed.

palisadoes commented:
Please fix the conflicting files

prayanshchh commented:
increasing coverage for setup, will try to get this done by today

prayanshchh commented:
1. sorry this has been a little slow (had exams)
2. I will improve the coverage tomorrow
3. if we need to integrate sample data as well, then we can create a script and extend setup's functionality even more, is it required?

palisadoes commented:
1. Sample data is being fixed in another PR.
2. Don't wait for that

palisadoes commented:
1. Sample data is being fixed in another PR.
2. Don't wait for that

prayanshchh commented:
does this look good @palisadoes ?

palisadoes reviewed (changes requested):
1. Remove. This is not relevant. Only production and development setups are required.
    ![image](https://github.com/user-attachments/assets/953ff269-4fc7-417e-979c-a9e4d2e935a4)
1. Why is this option necessary?
    ![image](https://github.com/user-attachments/assets/0ddd4e59-1d3f-4590-92cc-8a300628728a)
1. When the script is complete all files from the `envFiles/` directory are deleted.

prayanshchh commented:
> 1. Remove. This is not relevant. Only production and development setups are required.
>    ![image](https://private-user-images.githubusercontent.com/16875803/411260725-953ff269-4fc7-417e-979c-a9e4d2e935a4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkwNDEwNTcsIm5iZiI6MTczOTA0MDc1NywicGF0aCI6Ii8xNjg3NTgwMy80MTEyNjA3MjUtOTUzZmYyNjktNGZjNy00MTdlLTk3OWMtYTllNGQyZTkzNWE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA4VDE4NTIzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFlMTJjNTgzOWY3OTQxMzM2MjljODc0OGQwMzE4MDdlZTg1MWJjNDYyMzYzNmM1MDBlMWZmZjRiYTBjZDI4ZGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.NyE1kvriBXcxOuG_Um1BcPCoT4_o_ze_Z9uk1y9ewFg)
> 2. Why is this option necessary?
>    ![image](https://private-user-images.githubusercontent.com/16875803/411261159-0ddd4e59-1d3f-4590-92cc-8a300628728a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkwNDEwNTcsIm5iZiI6MTczOTA0MDc1NywicGF0aCI6Ii8xNjg3NTgwMy80MTEyNjExNTktMGRkZDRlNTktMWQzZi00NTkwLTkyY2MtOGEzMDA2Mjg3MjhhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA4VDE4NTIzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYyNDg4MDkxNzUyOTY1MTA2YjcyMDQ5ZGUwMmJjMTMwNmU1NmYwZWFlMTc1NDFjMmRkMDgwYzNiOWMzODRlN2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.KYZAoPGlvxkorUJF7YbH3tDAJC-FH3slEoR3N__J5Zg)
> 3. When the script is complete all files from the `envFiles/` directory are deleted.

set CI is needed, CI when true loads `env` as per `.env.ci` and when false sets up as per `.env.devcontainer`, makes it easy to setup testing or development setup 

I will remove Node enviorment

will implement 3

palisadoes commented:
Yes, but nobody is going to run setup interactively for CI. It's not necessary. CI should be only for the GitHub actions PR process. What other scenarios were you thinking about?

prayanshchh commented:
alright then I will remove CI part, maybe setting CI interactively isn't needed, also why should we delete files from `envFiles`, what if one has development setup and wants to switch to CI setup, having `.env.ci` in `envFiles` will help

prayanshchh commented:
> Yes, but nobody is going to run setup interactively for CI. It's not necessary. CI should be only for the GitHub actions PR process. What other scenarios were you thinking about?

 yes, but after removal contributors will have to copy `.env.ci `manually to `.env` which doesn't harm much, but this provides a more direct for the same.

palisadoes commented:
1. Your script deletes those files and they must remain. 
2. They are a part of the CI/CD pipeline
3. They are referred to frequently in the documentation
4. The issue did not require you to delete files that are a part of the repository. They must remain

prayanshchh commented:
[Screencast from 2025-02-09 00-50-43.webm](https://github.com/user-attachments/assets/935e389f-47a4-4f42-a366-116d02dfe0ff)

this video shows the working of script, no files in envFiles are getting deleted however in script .env.backup gets created just in case there is any error, this backup remains to store the previous config but this gets deleted if the setup is successful

prayanshchh commented on src/setup/setup.ts:
```diff
@@ -0,0 +1,519 @@
+
+	dotenv.config({ path: envFileName });
+
+	process.on("SIGINT", () => {
+		console.log("\nProcess interrupted! Undoing changes...");
+		answers = {};
+		if (fs.existsSync(".env.backup")) {
+			fs.copyFileSync(".env.backup", ".env");
+		}
+		process.exit(1);
+	});
```
SIGINT was done to handle ctrl + c behviour of the user, what will SIGTERM handle?

prayanshchh commented on src/setup/setup.ts:
```diff
@@ -0,0 +1,519 @@
+
+	await administratorEmail();
+
+	updateEnvVariable(answers);
+	console.log("Configuration complete.");
+	if (fs.existsSync(".env.backup")) {
+		fs.unlinkSync(".env.backup");
+	}
```
if backup is removed in between the script, then if there is some sort of failure there will no backup of the old env, hence when config is completed we remove the backup, else it persists

prayanshchh commented on test/setup/cloudbeaverSetup.test.ts:
```diff
@@ -0,0 +1,53 @@
+	});
+
+	it("should prompt the user for CloudBeaver configuration and update process.env", async () => {
+		const mockResponses = [
+			{ envReconfigure: true },
+			{ CI: "false" },
+			{ useDefaultApi: true },
+			{ useDefaultMinio: true },
+			{ useDefaultCloudbeaver: false },
+			{ CLOUDBEAVER_ADMIN_NAME: "mocked-admin" },
+			{ CLOUDBEAVER_ADMIN_PASSWORD: "mocked-password" },
+			{ CLOUDBEAVER_MAPPED_HOST_IP: "127.0.0.1" },
+			{ CLOUDBEAVER_MAPPED_PORT: "8080" },
+			{ CLOUDBEAVER_SERVER_NAME: "Mocked Server" },
+			{ CLOUDBEAVER_SERVER_URL: "https://127.0.0.1:8080" },
+			{ useDefaultPostgres: true },
+			{ API_ADMINISTRATOR_USER_EMAIL_ADDRESS: "test@email.com" },
+		];
```
in other test cases or files, validateEmail, validatePort is already tested, so doing it here again will be redundant, and if there is something cloudbeaver specific we can include it here, also for incomplete credentails can u also reccomend changes in implementation in setup.ts for cloudbeaver incomplete credentials and tests as well

prayanshchh commented on test/setup/postgresSetup.test.ts:
```diff
@@ -0,0 +1,76 @@
+		vi.resetAllMocks();
+	});
+
+	it("should prompt the user for Postgres configuration and update process.env", async () => {
+		const mockResponses = [
+			{ POSTGRES_DB: "mocked-db" },
+			{ POSTGRES_PASSWORD: "mocked-password" },
+			{ POSTGRES_USER: "mocked-user" },
+		];
+
+		const promptMock = vi.spyOn(inquirer, "prompt");
+		for (const response of mockResponses) {
+			promptMock.mockResolvedValueOnce(response);
+		}
+
+		const answers = await postgresSetup();
+
+		const expectedEnv = {
+			POSTGRES_DB: "mocked-db",
+			POSTGRES_PASSWORD: "mocked-password",
+			POSTGRES_USER: "mocked-user",
+		};
+
+		for (const [key, value] of Object.entries(expectedEnv)) {
+			expect(answers[key]).toBe(value);
+		}
+	});
```
in other test cases or files, validateEmail, validatePort is already tested, so doing it here again will be redundant, and if there is something postgres specific recomend changes in implementation in setup.ts for postgres incomplete credentials and tests as well

prayanshchh commented on test/setup/minioSetup.test.ts:
```diff
@@ -0,0 +1,81 @@
+		vi.resetAllMocks();
+	});
+
+	it("should prompt the user for Minio configuration and update process.env", async () => {
+		const mockResponses = [
+			{ MINIO_BROWSER: "off" },
+			{ MINIO_ROOT_PASSWORD: "mocked-password" },
+			{ MINIO_ROOT_USER: "mocked-user" },
+		];
+
+		const promptMock = vi.spyOn(inquirer, "prompt");
+
+		for (const response of mockResponses) {
+			promptMock.mockResolvedValueOnce(response);
+		}
+
+		const answers = await minioSetup();
+
+		const expectedEnv = {
+			MINIO_BROWSER: "off",
+			MINIO_ROOT_PASSWORD: "mocked-password",
+			MINIO_ROOT_USER: "mocked-user",
+		};
+
+		for (const [key, value] of Object.entries(expectedEnv)) {
+			expect(answers[key]).toBe(value);
+		}
+	});
```
in other test cases or files, validateEmail, validatePort is already tested, so doing it here again will be redundant

prayanshchh commented on test/setup/minioSetup.test.ts:
```diff
@@ -0,0 +1,81 @@
+
+	it("should prompt extended Minio config fields when CI=false", async () => {
+		const mockResponses = [
+			{ envReconfigure: true },
+			{ CI: "false" },
+			{ useDefaultApi: true },
+			{ useDefaultMinio: false },
+			{ MINIO_BROWSER: "on" },
+			{ MINIO_API_MAPPED_HOST_IP: "1.2.3.4" },
+			{ MINIO_API_MAPPED_PORT: "9000" },
+			{ MINIO_CONSOLE_MAPPED_HOST_IP: "1.2.3.5" },
+			{ MINIO_CONSOLE_MAPPED_PORT: "9001" },
+			{ MINIO_ROOT_PASSWORD: "mocked-password" },
+			{ MINIO_ROOT_USER: "mocked-user" },
+			{ useDefaultCloudbeaver: true },
+			{ useDefaultPostgres: true },
+			{ API_ADMINISTRATOR_USER_EMAIL_ADDRESS: "test@email.com" },
+		];
+
+		const promptMock = vi.spyOn(inquirer, "prompt");
+		for (const response of mockResponses) {
+			promptMock.mockResolvedValueOnce(response);
+		}
+
+		await setup();
+		dotenv.config({ path: ".env" });
+
+		const expectedEnv = {
+			MINIO_BROWSER: "on",
+			MINIO_API_MAPPED_HOST_IP: "1.2.3.4",
+			MINIO_CONSOLE_MAPPED_HOST_IP: "1.2.3.5",
+			MINIO_CONSOLE_MAPPED_PORT: "9001",
+			MINIO_ROOT_PASSWORD: "mocked-password",
+			MINIO_ROOT_USER: "mocked-user",
+		};
+
+		for (const [key, value] of Object.entries(expectedEnv)) {
+			expect(process.env[key]).toBe(value);
+		}
+	});
```
can u edit the implementation for this in setup, then give test according to that

prayanshchh commented on test/setup/setup.test.ts:
```diff
@@ -0,0 +1,157 @@
+		const promptMock = vi.spyOn(inquirer, "prompt");
+		for (const response of mockResponses) {
+			promptMock.mockResolvedValueOnce(response);
+		}
+
+		await setup();
+
+		const expectedEnv = {
+			API_BASE_URL: "http://127.0.0.1:4000",
+			API_HOST: "0.0.0.0",
+			API_PORT: "4000",
+			API_IS_APPLY_DRIZZLE_MIGRATIONS: "true",
+			API_JWT_EXPIRES_IN: "2592000000",
+			API_LOG_LEVEL: "info",
+			API_MINIO_ACCESS_KEY: "talawa",
+			API_MINIO_END_POINT: "minio",
+			API_MINIO_PORT: "9000",
+			API_MINIO_TEST_END_POINT: "minio-test",
+			API_MINIO_USE_SSL: "false",
+			API_POSTGRES_DATABASE: "talawa",
+			API_POSTGRES_HOST: "postgres",
+			API_POSTGRES_PORT: "5432",
+			API_POSTGRES_SSL_MODE: "false",
+			API_POSTGRES_TEST_HOST: "postgres-test",
+			API_POSTGRES_USER: "talawa",
+			CI: "false",
+			MINIO_ROOT_USER: "talawa",
+			CLOUDBEAVER_ADMIN_NAME: "talawa",
+			CLOUDBEAVER_MAPPED_HOST_IP: "127.0.0.1",
+			CLOUDBEAVER_MAPPED_PORT: "8978",
+			CLOUDBEAVER_SERVER_NAME: "Talawa CloudBeaver Server",
+			CLOUDBEAVER_SERVER_URL: "http://127.0.0.1:8978",
+		};
+
+		dotenv.config({ path: ".env" });
+
+		for (const [key, value] of Object.entries(expectedEnv)) {
+			expect(process.env[key]).toBe(value);
+		}
+	});
```
the expected env already has this
```
CLOUDBEAVER_ADMIN_NAME: "talawa",
CLOUDBEAVER_MAPPED_HOST_IP: "127.0.0.1",
CLOUDBEAVER_MAPPED_PORT: "8978",
CLOUDBEAVER_SERVER_NAME: "Talawa CloudBeaver Server",
CLOUDBEAVER_SERVER_URL: "http://127.0.0.1:8978",
```

what's the need for explicitly verifying cloudbeaver?

prayanshchh commented on test/setup/apiSetup.test.ts:
```diff
@@ -0,0 +1,163 @@
+		}
+	});
+});
+describe("validateURL", () => {
+	it("should return true for a valid URL", () => {
+		expect(validateURL("https://example.com")).toBe(true);
+		expect(validateURL("http://localhost:3000")).toBe(true);
+		expect(validateURL("ftp://ftp.example.com")).toBe(true);
+	});
+
+	it("should return an error message for an invalid URL", () => {
+		expect(validateURL("invalid-url")).toBe("Please enter a valid URL.");
+		expect(validateURL(" ")).toBe("Please enter a valid URL.");
+	});
+});
```
does the implementation support this, if not give the recommended implementation for validateURL

prayanshchh replied:
yes update test cases

prayanshchh commented on test/setup/apiSetup.test.ts:
```diff
@@ -0,0 +1,163 @@
+
+		const promptMock = vi.spyOn(inquirer, "prompt");
+
+		for (const response of mockResponses) {
+			promptMock.mockResolvedValueOnce(response);
+		}
+
+		const answers = await setup();
+		dotenv.config({ path: ".env" });
+
+		const expectedEnv = {
+			API_BASE_URL: "http://localhost:5000",
+			API_HOST: "127.0.0.1",
+			API_PORT: "5000",
+			API_IS_APPLY_DRIZZLE_MIGRATIONS: "true",
+			API_IS_GRAPHIQL: "false",
+			API_IS_PINO_PRETTY: "false",
+			API_JWT_EXPIRES_IN: "3600000",
+			API_JWT_SECRET: "mocked-secret",
+			API_LOG_LEVEL: "info",
+			API_MINIO_ACCESS_KEY: "mocked-access-key",
+			API_MINIO_END_POINT: "mocked-endpoint",
+			API_MINIO_PORT: "9001",
+			API_MINIO_SECRET_KEY: "mocked-secret-key",
+			API_MINIO_TEST_END_POINT: "mocked-test-endpoint",
+			API_MINIO_USE_SSL: "true",
+			API_POSTGRES_DATABASE: "mocked-database",
+			API_POSTGRES_HOST: "mocked-host",
+			API_POSTGRES_PASSWORD: "mocked-password",
+			API_POSTGRES_PORT: "5433",
+			API_POSTGRES_SSL_MODE: "true",
+			API_POSTGRES_TEST_HOST: "mocked-test-host",
+			API_POSTGRES_USER: "mocked-user",
+		};
+
+		for (const [key, value] of Object.entries(expectedEnv)) {
+			expect(answers[key]).toBe(value);
+		}
+	});
+});
```
we already have testing of validURL in apisetup, won't this be redundant?

prayanshchh commented:
1. setNodeEnviorment has been removed
2. @palisadoes does this look good now?

prayanshchh commented on src/setup/setup.ts:
```diff
@@ -0,0 +1,572 @@
+	return result;
+}
+
+let answers: Record<string, string> = {};
```
can you then provide a better approach then

prayanshchh commented on src/setup/setup.ts:
```diff
@@ -0,0 +1,572 @@
+	}
+}
+
+export async function setCI(): Promise<Record<string, string>> {
+	try {
+		answers.CI = await promptList("CI", "Set CI:", ["true", "false"], "false");
+	} catch (err) {
+		console.error(err);
+		if (fs.existsSync(".env.backup")) {
+			fs.copyFileSync(".env.backup", ".env");
+		}
+		process.exit(1);
+	}
+	return answers;
+}
+
+export async function administratorEmail(): Promise<Record<string, string>> {
+	try {
+		answers.API_ADMINISTRATOR_USER_EMAIL_ADDRESS = await promptInput(
+			"API_ADMINISTRATOR_USER_EMAIL_ADDRESS",
+			"Enter email:",
+			"administrator@email.com",
+			validateEmail,
+		);
+	} catch (err) {
+		console.log(err);
+		if (fs.existsSync(".env.backup")) {
+			fs.copyFileSync(".env.backup", ".env");
+		}
+		process.exit(1);
+	}
+	return answers;
+}
```
please provide the recommended approach then

prayanshchh commented on src/setup/setup.ts:
```diff
@@ -0,0 +1,590 @@
+		answers = await postgresSetup(answers);
+	}
+
+	answers = await administratorEmail(answers);
+
+	updateEnvVariable(answers);
+	console.log("Configuration complete.");
+	if (fs.existsSync(".env.backup")) {
+		fs.unlinkSync(".env.backup");
+	}
+	return answers;
```
if everything goes right we don't need backup, that's why we delete it in the end

prayanshchh commented on src/setup/setup.ts:
```diff
@@ -0,0 +1,590 @@
+	}
+}
+
+export async function setCI(answers: SetupAnswers): Promise<SetupAnswers> {
+	try {
+		answers.CI = await promptList("CI", "Set CI:", ["true", "false"], "false");
+	} catch (err) {
+		console.error(err);
+		if (fs.existsSync(".env.backup")) {
+			fs.copyFileSync(".env.backup", ".env");
+		}
+		process.exit(1);
+	}
+	return answers;
+}
```
can u create the function for that

prayanshchh commented on test/setup/minioSetup.test.ts:
```diff
@@ -0,0 +1,101 @@
+		}
+	});
+	it("should detect port conflicts between API and Console ports", async () => {
+		const inputAnswers = { CI: "false" };
+
+		const mockResponses = [
+			{ MINIO_BROWSER: "on" },
+			{ MINIO_API_MAPPED_HOST_IP: "127.0.0.1" },
+			{ MINIO_API_MAPPED_PORT: "9000" },
+			{ MINIO_CONSOLE_MAPPED_HOST_IP: "127.0.0.1" },
+			{ MINIO_CONSOLE_MAPPED_PORT: "9000" },
+		];
+
+		const promptMock = vi.spyOn(inquirer, "prompt");
+		for (const response of mockResponses) {
+			promptMock.mockResolvedValueOnce(response);
+		}
+
+		await expect(minioSetup(inputAnswers)).rejects.toThrow(
+			"Port conflict detected: MinIO API and Console ports must be different",
+		);
+	});
```
we already check CI = true in setup.test.ts, won't it be redundant?

prayanshchh commented on src/setup/setup.ts:
```diff
@@ -0,0 +1,587 @@
+import inquirer from "inquirer";
+import { updateEnvVariable } from "./updateEnvVariable";
+
+interface SetupAnswers {
+	[key: string]: string;
+}
```
can u create this for all variables

prayanshchh replied:
should we write this logic in another file

prayanshchh replied:
I don't think we should do this as in tests we don't pass the answers with all the values, and creating new types for every test doesn't make sense as it would be redundant

prayanshchh commented on src/setup/setup.ts:
```diff
@@ -0,0 +1,587 @@
+
+	answers.API_POSTGRES_HOST = await promptInput(
+		"API_POSTGRES_HOST",
+		"Postgres host:",
+		"postgres",
+	);
+
+	answers.API_POSTGRES_PASSWORD = await promptInput(
+		"API_POSTGRES_PASSWORD",
+		"Postgres password:",
+		"password",
+	);
+
+	answers.API_POSTGRES_PORT = await promptInput(
+		"API_POSTGRES_PORT",
+		"Postgres port:",
+		"5432",
+	);
+
+	answers.API_POSTGRES_SSL_MODE = await promptList(
+		"API_POSTGRES_SSL_MODE",
+		"Use Postgres SSL?",
+		["true", "false"],
+		"false",
+	);
+
+	answers.API_POSTGRES_TEST_HOST = await promptInput(
+		"API_POSTGRES_TEST_HOST",
+		"Postgres test host:",
+		"postgres-test",
+	);
+
+	answers.API_POSTGRES_USER = await promptInput(
+		"API_POSTGRES_USER",
+		"Postgres user:",
+		"talawa",
+	);
+
+	return answers;
+}
```
can u write test for this in setup.test.ts to maintain coverage

prayanshchh commented on src/setup/setup.ts:
```diff
@@ -0,0 +1,591 @@
+}
+
+export async function postgresSetup(
+	answers: SetupAnswers,
+): Promise<SetupAnswers> {
+	answers.POSTGRES_DB = await promptInput(
+		"POSTGRES_DB",
+		"Postgres database:",
+		"talawa",
+	);
+
+	if (answers.CI === "false") {
+		answers.POSTGRES_MAPPED_HOST_IP = await promptInput(
+			"POSTGRES_MAPPED_HOST_IP",
+			"Postgres mapped host IP:",
+			"127.0.0.1",
+		);
+
+		answers.POSTGRES_MAPPED_PORT = await promptInput(
+			"POSTGRES_MAPPED_PORT",
+			"Postgres mapped port:",
+			"5432",
+			validatePort,
+		);
+	}
+
+	answers.POSTGRES_PASSWORD = await promptInput(
+		"POSTGRES_PASSWORD",
+		"Postgres password:",
+		"password",
+	);
+
+	answers.POSTGRES_USER = await promptInput(
+		"POSTGRES_USER",
+		"Postgres user:",
+		"talawa",
+	);
+
+	return answers;
+}
```
can u create test for this

prayanshchh commented on src/setup/setup.ts:
```diff
@@ -0,0 +1,591 @@
+		"CloudBeaver admin name:",
+		"talawa",
+		validateCloudBeaverAdmin,
+	);
+
+	answers.CLOUDBEAVER_ADMIN_PASSWORD = await promptInput(
+		"CLOUDBEAVER_ADMIN_PASSWORD",
+		"CloudBeaver admin password:",
+		"password",
+		validateCloudBeaverPassword,
+	);
+
+	answers.CLOUDBEAVER_MAPPED_HOST_IP = await promptInput(
+		"CLOUDBEAVER_MAPPED_HOST_IP",
+		"CloudBeaver mapped host IP:",
+		"127.0.0.1",
+	);
+
+	answers.CLOUDBEAVER_MAPPED_PORT = await promptInput(
+		"CLOUDBEAVER_MAPPED_PORT",
+		"CloudBeaver mapped port:",
+		"8978",
+		validatePort,
+	);
+
+	answers.CLOUDBEAVER_SERVER_NAME = await promptInput(
+		"CLOUDBEAVER_SERVER_NAME",
+		"CloudBeaver server name:",
+		"Talawa CloudBeaver Server",
+	);
+
+	answers.CLOUDBEAVER_SERVER_URL = await promptInput(
+		"CLOUDBEAVER_SERVER_URL",
+		"CloudBeaver server URL:",
+		"http://127.0.0.1:8978",
+		validateCloudBeaverURL,
+	);
+
+	return answers;
+}
```
can u create test for this

palisadoes reviewed (changes requested):
1. Please make coderabbit.ai approves your work.
2. Make sure all tests pass and are valid.
3. Ensure the test code coverage for your patch reaches as close to 100% as possible.

prayanshchh commented:
will try to get this by today

prayanshchh commented on test/setup/apiSetup.test.ts:
```diff
@@ -0,0 +1,215 @@
+	});
+});
+describe("validateURL", () => {
+	it("should validate standard URLs", () => {
+		expect(validateURL("https://example.com")).toBe(true);
+		expect(validateURL("http://localhost:3000")).toBe(true);
+	});
+
+	it("should validate URLs with complex components", () => {
+		expect(validateURL("https://sub.example.com:8080/path")).toBe(true);
+		expect(validateURL("http://sub.domain.example.com/path?query=1")).toBe(
+			true,
+		);
+		expect(validateURL("https://example.com/path#fragment")).toBe(true);
+	});
+
+	it("should validate IP addresses", () => {
+		expect(validateURL("http://127.0.0.1:4000")).toBe(true);
+		expect(validateURL("https://[::1]:8080")).toBe(true);
+		expect(validateURL("http://192.168.1.1")).toBe(true);
+	});
+
+	it("should reject invalid protocols", () => {
+		expect(validateURL("ftp://example.com")).toBe(
+			"Please enter a valid URL with http:// or https:// protocol.",
+		);
+		expect(validateURL("ws://example.com")).toBe(
+			"Please enter a valid URL with http:// or https:// protocol.",
+		);
+	});
+
+	it("should reject malformed URLs", () => {
+		expect(validateURL("http://")).toBe("Please enter a valid URL.");
+		expect(validateURL("http://example.com:abc")).toBe(
+			"Please enter a valid URL.",
+		);
+		expect(validateURL(" ")).toBe("Please enter a valid URL.");
+		expect(validateURL("")).toBe("Please enter a valid URL.");
+	});
+});
```
can u create test for this, and also check if implementation supports validation for case sensitive URL's

prayanshchh replied:
yes, fix implementation

prayanshchh commented on test/setup/apiSetup.test.ts:
```diff
@@ -0,0 +1,215 @@
+		consoleErrorSpy.mockRestore();
+	});
+
+	it("should handle prompt errors correctly", async () => {
+		const processExitSpy = vi
+			.spyOn(process, "exit")
+			.mockImplementation(() => undefined as never);
+		const fsExistsSyncSpy = vi.spyOn(fs, "existsSync").mockReturnValue(true);
+		const fsCopyFileSyncSpy = vi
+			.spyOn(fs, "copyFileSync")
+			.mockImplementation(() => undefined);
+
+		const mockError = new Error("Prompt failed");
+		vi.spyOn(inquirer, "prompt").mockRejectedValueOnce(mockError);
+
+		const consoleErrorSpy = vi.spyOn(console, "error");
+
+		await apiSetup({});
+
+		expect(consoleErrorSpy).toHaveBeenCalledWith(mockError);
+		expect(fsExistsSyncSpy).toHaveBeenCalledWith(".env.backup");
+		expect(fsCopyFileSyncSpy).toHaveBeenCalledWith(".env.backup", ".env");
+		expect(processExitSpy).toHaveBeenCalledWith(1);
+
+		vi.clearAllMocks();
+	});
+});
```
can you create those tests and also check if implentation supports that.

prayanshchh commented on src/setup/setup.ts:
```diff
@@ -0,0 +1,591 @@
+	}
+}
+
+function handlePromptError(err: unknown): never {
+	console.error(err);
+	if (fs.existsSync(".env.backup")) {
+		fs.copyFileSync(".env.backup", ".env");
+	}
+	process.exit(1);
+}
```
what should be approrpiate then, throwing the err?

prayanshchh replied:
I think we should do process.exit, it gives a clean exit from errors, errors might be unpredictable it's better if the contributor restarts the setup instead of continuing with erroneous setup

prayanshchh commented:
1. All suggestions by coderrabit have been implemented, some seemed unnecessary so have been ignored
2. coverage is 95+ now
3. this has been tested thoroughly, happy paths and unpredictable inputs.
4. can you recommend changes or if it looks good merge this.

prayanshchh commented:
@palisadoes can you have a look at this

palisadoes reviewed (changes requested):
1. Please add documentation on how run the script in `docs/docs/docs/getting-started/installation.md`
2. I'd like to merge this very soon.

palisadoes commented:
In the file in my last comment.

prayanshchh commented:
I meant the setup folder which currently exists in `src/setup`, previously we didn't have `scripts/` but now we do.

palisadoes commented:
Yes in a `scripts/setup` folder.

prayanshchh commented on docs/docs/docs/getting-started/installation.md:
````diff
@@ -41,6 +41,23 @@
+pnpm tsx setup.ts
+```
+
+The script will ask whether you’re in CI mode (CI=true) or non-CI (CI=false) mode. Choose:
+
+1. CI=false for local/development environments.
+2. CI=true for testing or continuous integration pipelines.
+3. It will also ask whether you want to use recommended defaults. Answer “yes” to quickly accept safe defaults or “no” to provide custom inputs. Once the prompts finish, your .env file will be generated or updated.
````
can u add this

prayanshchh commented on scripts/setup/setup.ts:
```diff
@@ -0,0 +1,591 @@
+			answers.CI === "true" ? "off" : "on",
+		);
+
+		if (answers.CI === "false") {
+			answers.MINIO_API_MAPPED_HOST_IP = await promptInput(
+				"MINIO_API_MAPPED_HOST_IP",
+				"Minio API mapped host IP:",
+				"127.0.0.1",
+			);
+
+			answers.MINIO_API_MAPPED_PORT = await promptInput(
+				"MINIO_API_MAPPED_PORT",
+				"Minio API mapped port:",
+				"9000",
+				validatePort,
+			);
+
+			answers.MINIO_CONSOLE_MAPPED_HOST_IP = await promptInput(
+				"MINIO_CONSOLE_MAPPED_HOST_IP",
+				"Minio console mapped host IP:",
+				"127.0.0.1",
+			);
+
+			answers.MINIO_CONSOLE_MAPPED_PORT = await promptInput(
+				"MINIO_CONSOLE_MAPPED_PORT",
+				"Minio console mapped port:",
+				"9001",
+				validatePort,
+			);
+
+			if (answers.MINIO_API_MAPPED_PORT === answers.MINIO_CONSOLE_MAPPED_PORT) {
+				throw new Error(
+					"Port conflict detected: MinIO API and Console ports must be different",
+				);
+			}
+		}
```
can u implement this

prayanshchh replied:
also add tests for ur changes in mino setup tests

prayanshchh replied:
also add tests for ur changes in mino setup tests

prayanshchh commented on test/setup/apiSetup.test.ts:
```diff
@@ -0,0 +1,285 @@
+
+		const promptMock = vi.spyOn(inquirer, "prompt");
+
+		for (const response of mockResponses) {
+			promptMock.mockResolvedValueOnce(response);
+		}
+
+		const answers = await setup();
+		dotenv.config({ path: ".env" });
+
+		const expectedEnv = {
+			API_BASE_URL: "http://localhost:5000",
+			API_HOST: "127.0.0.1",
+			API_PORT: "5000",
+			API_IS_APPLY_DRIZZLE_MIGRATIONS: "true",
+			API_IS_GRAPHIQL: "false",
+			API_IS_PINO_PRETTY: "false",
+			API_JWT_EXPIRES_IN: "3600000",
+			API_JWT_SECRET: "mocked-secret",
+			API_LOG_LEVEL: "info",
+			API_MINIO_ACCESS_KEY: "mocked-access-key",
+			API_MINIO_END_POINT: "mocked-endpoint",
+			API_MINIO_PORT: "9001",
+			API_MINIO_SECRET_KEY: "mocked-secret-key",
+			API_MINIO_TEST_END_POINT: "mocked-test-endpoint",
+			API_MINIO_USE_SSL: "true",
+			API_POSTGRES_DATABASE: "mocked-database",
+			API_POSTGRES_HOST: "mocked-host",
+			API_POSTGRES_PASSWORD: "mocked-password",
+			API_POSTGRES_PORT: "5433",
+			API_POSTGRES_SSL_MODE: "true",
+			API_POSTGRES_TEST_HOST: "mocked-test-host",
+			API_POSTGRES_USER: "mocked-user",
+			API_ADMINISTRATOR_USER_EMAIL_ADDRESS: "test@email.com",
+		};
+
+		for (const [key, value] of Object.entries(expectedEnv)) {
+			expect(answers[key]).toBe(value);
+		}
+	});
```
we already have tested defualts, if base url or port is missing defaults will be used

prayanshchh commented on test/setup/cloudbeaverSetup.test.ts:
```diff
@@ -0,0 +1,127 @@
+		});
+	});
+
+	describe("validateCloudBeaverURL", () => {
+		it("should validate server URL format", () => {
+			expect(validateCloudBeaverURL("")).toBe("Server URL is required");
+			expect(validateCloudBeaverURL("invalid")).toBe("Invalid URL format");
+			expect(validateCloudBeaverURL("ftp://127.0.0.1")).toBe(
+				"URL must use HTTP or HTTPS protocol",
+			);
+			expect(validateCloudBeaverURL("http://127.0.0.1:8978")).toBe(true);
+			expect(validateCloudBeaverURL("https://localhost:8978")).toBe(true);
+		});
+	});
```
Node’s URL constructor rejects invalid port numbers (like 99999) by throwing a TypeError (or similar) right away. hence  in implentation we should remove port check and the test should expect  Invalid URL format

prayanshchh commented on scripts/setup/updateEnvVariable.ts:
```diff
@@ -0,0 +1,48 @@
+
+		let updatedContent: string = existingContent;
+
+		for (const key in config) {
+			const value = config[key];
+			const regex = new RegExp(
+				`^${key.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}=.*`,
+				"gm",
+			);
+
+			if (regex.test(updatedContent)) {
+				updatedContent = updatedContent.replace(regex, `${key}=${value}`);
+			} else {
+				updatedContent += `\n${key}=${value}`;
+			}
```
do we need to add test for this new function?

prayanshchh replied:
add tests for this then in updateenvvariable.spec.ts

prayanshchh commented on docs/docs/docs/getting-started/installation.md:
````diff
@@ -41,6 +41,35 @@ You will need to configure the API to work correctly.
+## Running the Setup Script
+To configure the .env file, run one of the following commands in your project’s root directory:
+
+```bash
+npm run setup
+```
````
## Running the Setup Script
To configure the .env file, run one of the following commands in your project’s root directory:


```bash
npm run setup
```

or

```bash
pnpm tsx setup.ts
```

The script will ask whether you're in CI mode (CI=true) or non-CI (CI=false) mode. Choose:


is this good?

prayanshchh commented:
documentation on how to run the script has been added

palisadoes commented:
@prayanshchh These are the mandatory parameters. Some need to match.

https://docs-api.talawa.io/docs/installation#production-environment-setup

prayanshchh commented:
will make the changes, working on a priority issue, please consider

prayanshchh commented on test/scripts/setup/caddySetup.test.ts:
```diff
@@ -0,0 +1,49 @@
+		const answers = await caddySetup({});
+		dotenv.config({ path: ".env" });
+
+		const expectedEnv = {
+			CADDY_HTTP_MAPPED_PORT: "801",
+			CADDY_HTTPS_MAPPED_PORT: "4431",
+			CADDY_HTTP3_MAPPED_PORT: "4431",
+			CADDY_TALAWA_API_DOMAIN_NAME: "localhost",
+			CADDY_TALAWA_API_EMAIL: "example@talawa.com",
+			CADDY_TALAWA_API_HOST: "api",
+			CADDY_TALAWA_API_PORT: "3000",
+		};
+
+		for (const [key, value] of Object.entries(expectedEnv)) {
+			expect(answers[key]).toBe(value);
+		}
+	});
+});
```
we already have vaildate port and validate email checks, this will be redundant

prayanshchh commented:
@tasneemkoushar @meetulr this is ready

prayanshchh commented:
@palisadoes PTAL

prayanshchh commented:
this has too many commits due to conflicts, can we merge this now?
